### PR TITLE
[core/loader] Fix presharded cache-key gaps: moe_dense_tp_size, EPLB with structural signature

### PR DESCRIPTION
**Pull request by shuwang21**
## Motivation

`PreshardedModelLoader` caches per-rank weight shards under a subfolder derived from the sharding config. If two launches with different configs map to the same subfolder, the loader silently serves wrong weights.

Two concrete gaps in the existing cache key:

- **`moe_dense_tp_size`**: controls dense-layer TP width independently of global `tp_size`. A run with `--moe-dense-tp-size 1` and a default run both produced `TP-4`, colliding on the same directory.
- **EPLB settings** (`ep_num_redundant_experts`, `init_expert_location`): change which logical expert's weights land in each physical slot without changing tensor shapes, so they must be explicit key fields.

More broadly, every new sharding dimension requires a manual update to the key-building code — easy to forget, and the subfolder name grows with each addition. This PR restructures the key as `TP-{tp}-sig-{hash16}` and adds a structural signature as a self-correcting safety net.

## Modifications

**`loader.py` — `_collect_shard_config` (new) + `_build_subfolder_name` (restructured)**

All key dimensions (TP/DP/EP/PP, `moe_dense_tp_size`, quantization, model dtype, EPLB fields, structural signature) are collected into a JSON-native dict and folded into one SHA1 hash over its canonical encoding. The subfolder name is now `TP-{tp}-sig-{hash16}` — TP stays human-readable, the name never grows, and adding a future dimension only requires extending `_collect_shard_config`.

**`loader.py` — shard config persisted and verified**

The full shard config dict is written into `checksum.json` under the `shard_config` key (atomically with the dump plan, by rank 0), serving as human-readable provenance for the hashed name. At load time it is compared field-by-field against the current config; a mismatch (only possible via hash collision or a key-derivation bug) logs both configs and degrades to a cache miss with re-dump — never a silent wrong-weights load.

**`loader.py` — `_compute_structural_signature` / `_hash_structural_signature` (new)**

Constructs a meta-device model under the live parallel state and hashes `sorted((name, shape, dtype))` over its `state_dict()`. Since `nn.Module.__init__` reads live TP/EP/PP world sizes to size each parameter, any future sharding knob that changes per-rank shapes automatically changes the hash — no manual update needed. Falls back to `structural_signature: null` in the key if meta-device construction fails, leaving the manually enumerated fields as the only protection (logged as a warning).

`state_dict()` covers all `nn.Parameter` and registered buffers but not extra tensors derived post-load (e.g. `self_attn.w_kc` in DeepSeek MLA, which is `None` at `__init__` time and only materialized during `load_weights()`). These are intentionally excluded: their shapes are fully determined by upstream parameters already in `state_dict()`, so any sharding change that would affect them is already reflected in the hash. Post-load correctness is covered by the existing `rank_checksum` verification.

**`loader.py` — fix meta-device pollution of the global rotary cache**

`get_rope()` caches rotary modules in a process-global `_ROPE_DICT` whose key does not include device. The meta-device pass would otherwise leave meta rotary modules in the cache for the real model init to reuse, failing with "Cannot copy out of meta tensor". `_compute_structural_signature` now removes meta-device entries from `_ROPE_DICT` in a `finally` block, so cleanup runs even when model construction fails midway.

**`test_presharded_loader.py`**

Added `TestStructuralSignature` (hash determinism, sensitivity to shape/dtype/name changes, failure path, `_ROPE_DICT` cleanup on failure, end-to-end with patched `_initialize_model`) and `TestShardConfig` (every enumerated field feeds the subfolder hash; load-time verification degrades to cache miss on mismatch/missing config, never raises).

## Accuracy Tests

No forward code changed. Existing `rank_checksum` verification remains the post-load correctness safety net.

## Speed Tests and Profiling

`_compute_structural_signature` runs once at startup on a meta-device model (no real memory allocation). No inference throughput impact.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

File: python/sglang/srt/model_loader/loader.py [hunk 1]
```diff
@@ -1630,10 +1634,69 @@ def _safe(fn) -> int:
...
+        if local_dense_tp != tp:
+            parts.append(f"DenseTP-{local_dense_tp}")
         if model_config.quantization:
             parts.append(f"dtype-{model_config.quantization}")
+        if ep_num_redundant_experts:
+            parts.append(f"RedEP-{ep_num_redundant_experts}")
+        if init_expert_location and init_expert_location != "trivial":
+            loc_hash = hashlib.sha1(
+                str(init_expert_location).encode()
+            ).hexdigest()[:8]
```
[the remark this answers is not in the document]
Comment (shuwang21) [reply python/sglang/srt/model_loader/loader.py]: Yes. This is possible. But in this case, we fall back to do rank_checksum for post correctness check.

File: python/sglang/srt/model_loader/loader.py [hunk 2]
```diff
@@ -1630,10 +1634,69 @@ def _safe(fn) -> int:
...
+    def _compute_structural_signature(
+        self, model_config: ModelConfig
+    ) -> Optional[str]:
+        try:
+            quant_config = _get_quantization_config(model_config, self.load_config)
+            with set_default_torch_dtype(model_config.dtype):
+                with torch.device("meta"):
+                    meta_model = _initialize_model(
+                        model_config, self.load_config, quant_config
+                    )
```
[the remark this answers is not in the document]
Comment (shuwang21) [reply python/sglang/srt/model_loader/loader.py:1722]: nice catch and explaination. Just fixed.

**Comment by shuwang21** [comment]
Discussed offline with @ch-wan: keep the presharded path simple as `TP-{tp}-sig-{hash16}`, where the hash covers all key dimensions `(parallel sizes, dtype, quantization, EPLB fields, structural signature)`. The whole shard config is persisted in checksum.json for debugging and verified at load time.
